sg_ident.8
.TH SG_IDENT "8" "January 2007" "sg3_utils\-1.23" SG3_UTILS
.SH NAME
sg_ident \- sends a SCSI REPORT or SET IDENTIFYING INFORMATION command
.SH SYNOPSIS
.B sg_ident
[\fI\-\-ascii\fR] [\fI\-\-clear\fR] [\fI\-\-help\fR] [\fI\-\-itype=IT\fR]
[\fI\-\-raw\fR] [\fI\-\-set\fR] [\fI\-\-verbose\fR] [\fI\-\-version\fR]
\fIDEVICE\fR
.SH DESCRIPTION
.\" Add any additional description here
.PP
Send a SCSI REPORT IDENTIFYING INFORMATION or SET IDENTIFYING INFORMATION
command to \fIDEVICE\fR. Prior to SPC\-4 (revison 7) these
commands were called REPORT DEVICE IDENTIFIER and SET DEVICE IDENTIFIER
respectively. SCSI devices that support these two commands allow users
to write (set) identifying information and report it back at some
later time. The information is persistent (i.e. stored on some
non\-volatile medium within the SCSI device that will survive a power
outage).
.PP
Typically the space allocated for the information is limited:
SPC\-4 (revision 7) states that for information type 0, the minimum
length is 64 bytes and the maximum is 512 bytes. For other information
types (1 to 126 inclusive) the maximum length is 256 bytes. Also
information types 1 to 126 (inclusive) should contain a null
terminated UTF\-8 string. The author has seen older disks that only
support 16 bytes.
.PP
The default action when no options are given is to invoke the
Report Identifying Information command with the information type defaulting
to zero. Error reports are sent to stderr. By default the information is
shown in ASCII\-HEX (up to 16 bytes per line) with an ASCII representation
to the right with dots replacing non printable characters.
.SH OPTIONS
Arguments to long options are mandatory for short options as well.
.TP
\fB\-A\fR, \fB\-\-ascii\fR
invokes the Report Identifying Information command and if anything is
found interprets it as ASCII (or UTF\-8 depending on the locale)
and prints the information to stdout.
.TP
\fB\-C\fR, \fB\-\-clear\fR
invokes the Set Identifying Information command with an information length
of zero. This has the effect of clearing the existing information.
.TP
\fB\-h\fR, \fB\-\-help\fR
output the usage message then exit.
.TP
\fB\-i\fR, \fB\-\-itype\fR=\fIIT\fR
where \fIIT\fR is the information type. Defaults to zero. The maximum value
is 127 which is special and cannot be used with \fI\-\-set\fR or
\fI\-\-clear\fR. The information type of 127 (if supported) causes the REPORT
IDENTIFYING INFORMATION command to respond with a list of available
information types and their maximum lengths in bytes. The odd numbered
information types between 3 and 125 (inclusive) are not to be used (as they
clash with the SCC\-2 standard).
.TP
\fB\-r\fR, \fB\-\-raw\fR
invokes the Report Identifying information command and if anything
is found sends the information (which may be binary) to stdout. Nothing else
is sent to stdout however error reports, if any, are sent to stderr.
.TP
\fB\-S\fR, \fB\-\-set\fR
first reads stdin until an EOF is detected then invokes the Set Identifying
Information command to set what has been fetched from stdin as the
information. The amount of data read must be between 1 and 512 bytes
length (inclusive).
.TP
\fB\-v\fR, \fB\-\-verbose\fR
increase the level of verbosity, (i.e. debug output).
.TP
\fB\-V\fR, \fB\-\-version\fR
print the version string and then exit.
.PP
This utility permits users to write their own identifying information to
their SCSI devices. There are several other types of descriptors (or
designators) that the user cannot change. These include the SCSI INQUIRY
command with its standard vendor and product identification strings and the
product revision level; plus the large amount of information provided by
the "Device Identification" VPD page (see sg_vpd). There is also the READ
MEDIA SERIAL NUMBER command (see sg_rmsn). The MMC\-4 command set for CD
and DVDs has a "media serial number" feature (0x109) [and a "logical unit
serial number" feature]. These can be viewed with the sg_get_config utility.
.SH EXAMPLES
First, to see if there is an existing information whose format
is unknown (for information type 0), use no options:
.PP
  # sg_ident /dev/sdb
.br
   00     31 32 33 34 35 36 37 38  39 30          1234567890
.PP
If it is ASCII then it can printed as such:
.PP
  # sg_ident \-\-ascii /dev/sdb
.br
  1234567890
.PP
The information can be copied to a file, cleared and then 
re\-asserted with this sequence:
.PP
  # sg_ident \-\-raw /dev/sdb > t
.br
  # sg_ident \-\-clear /dev/sdb
.br
  # cat t | sg_ident \-\-set /dev/sdb
.SH EXIT STATUS
The exit status of sg_ident is 0 when it is successful. Otherwise see
the sg3_utils(8) man page.
.SH AUTHORS
Written by Douglas Gilbert.
.SH "REPORTING BUGS"
Report bugs to <dgilbert at interlog dot com>.
.SH COPYRIGHT
Copyright \(co 2005\-2007 Douglas Gilbert
.br
This software is distributed under a FreeBSD license. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
.B sg_vpd(sg3_utils), sg_rmsn(sg3_utils), sg_get_config(sg3_utils)


CHANGELOG
Each utility has its own version number, date of last change and
some description at the top of its ".c" file. All utilities in the main
directory have their own "man" pages. There is also a sg3_utils man page.

Changelog for sg3_utils-1.25 [20070629]
  - sg_dd: add oflag=sparse to step over bs*bpt number of zeros;
    - with oflag=sparse, write last bs*bpt segment at end or after
      error so file length of OFILE is appropriate
    - allow iflag=null and oflag=null which do nothing (placeholders)
    - when coe>1 then SCSI READ LONG logic remember extended block
      length of first encountered error
  - sg_ses: sync with ses2r17

Changelog for sg3_utils-1.24 [20070507]
  - sg_raw: new utility to send arbirary SCSI commands
  - sg_luns: increase number of luns that can be fetched
    - fix length of raw and hex output
    - add '--quiet' option to output only ASCII hex
      representation of each lun
  - sg_rtpg: update for changes in spc4r09
  - sg_persist: update documentation, spc-4 references
    - fix exit status values
  - sg_inq: update version descriptors per spc4r09
    - fix '--id' and '--extended'
    - extend block limits VPD page (sbc3r09)
  - sg_vpd: extend block limits VPD page (sbc3r09)
    - append relative target port identifier to SAS target
      port address with '-iq' option
  - sg_logs: add decoding for stats+performance log pages
    - fix showing of page names for pdt > 0
    - implement '-HH' for single and all pages, fix '-r'
    - when '--maxlen=' given, only do single fetch
    - add Tape Alert (ssc), Media and Element statistics (smc) pages 
    - add '--brief' option
  - sg_ses: sync with ses2r16
    - fix bay number for SAS
  - sg_format: add '--dcrt' and '--security' options
  - sgm_dd: add 'smmap' oflag for shared_mmap_io testing
    - add 'dio' oflag
  - sg_dd, sgp_dd: add 'dio' iflag and oflag
  - sg_modes: change SAS mode page names per sas2r09
    - check validity of block descriptors length
  - sg_pt: change opaque context object from 'void *'
    to 'struct sg_pt_base *'
  - sg_opcodes: anticipate extra tmfs from 07-159r0
  - sg_sat_set_features: add more usage information
    - add man page
  - sg_sat_phy_event: add to examples directory
  - sg_lib: sync asc/ascq codes with spc4r10
  - Solaris port: using uscsi interface

Changelog for sg3_utils-1.23 [20070131]
  - sg_read_buffer: new utility
  - sg_write_buffer: new utility
  - sg_opcodes, sg_senddiag, sg_logs, sg_modes, sg_start, sg_inq,
    sg_turs, sg_readcap, sg_rbuf: add getopt_long() based cli;
    old and new cli selectable, new getopt_long cli is default
  - scripts: new subdirectory containing some bash scripts
    - add scripts/README file
  - sg_reassign: add '--hex' option for grown and primary lists
  - sg_rtpg: add '--raw' option
  - sg_lib.h, sg_cmds_basic.h + sg_cmds_extra.h: add C++ 
    'extern "C" ' wrappers
    - cleanup C code so it will compile as C++
  - sg_lib: sync with spc4r08
    - include <inttypes.h>, use PRId64 instead of %lld form
    - fix sg_get_sense_str() when empty sense buffer
  - win32 port: add Makefile.mingw + related support for MinGW
  - sg_cmds_extra: add sg_ll_read_buffer() and sg_ll_write_buffer()
  - sg_dd, sgp_dd, sgm_dd, sg_read: use lseek64() instead of llseek.c
  - sgm_dd: accept coe=<n> for interworking with sg_dd
  - sg_rdac: fix on non-linux ports
  - sg_ses: fix spurious warning in additional element status page
    - '-rr' option outputs a diagnostic page in binary to stdout
  - sg_opcodes: add command timeout descriptor support (spc4r08)
    - change linux specific pass through to generic pass through
  - sg_logs: add 'name=value' decoding for SAS specific lpage
  - examples+utils subdirectories: remove symlinks
  - synchronize man pages with usage messages
  - sg3_utils.spec: rework

Changelog for sg3_utils-1.22 [20061016]
  - sgp_dd: accept verbose=<n> as well as deb=<n> to ease
    interworking with sg_dd and sgm_dd
  - sg_sat_set_features: added to examples directory
  - sg_lib: sync asc/ascq text with spc4r06
    - move SG_LIB_CAT_NO_SENSE and SG_LIB_CAT_RECOVERED to
      20 and 21 respectively; add SG_LIB_CAT_ABORTED_COMMAND
      at 11 (its sense key value) 
  - sg_vpd: tweak '--page=sp --quiet' output
    - change '-HHH' so same as '-rr' (prepares ATA Information
      (ai) response for hdparm)
  - sg_requests: add '-s' option to set exit status from
    parameter data
  - sg_modes: exit quickly from '-e' if device not ready
  - sg_logs: sync sas log pages with sas2r05a
    - expand background scan results log page
    - add '-m=<max_len>' to limit response length
    - drop '-scum' and '-sthr' options and add '-select'
  - sg_write_long: add '--16' option to send 16 byte cdb
    - add '--wr_uncor' and '--pblock' options
  - sg_senddiag: cleanup and add sdiag_sas_p1_stop.txt
    to examples directory
  - sg_format: add '--cmplst=<n>' option (default: 1)
    - add '--pfu=<n>' option
    - expand man page to discuss P/D/C/GLISTs
  - sg_reassign: add '--primary' option to fetch primary
    defect list (PLIST) length
  - sg_readcap: add '-H' option to output response in hex
    and '-r' to output response in binary to stdout
    - add logical blocks per physical block (sbc3r07)
  - sginfo: add PLIST and GLIST designation to defect lists
  - sg_cmds: split this support file into sg_cmds_basic.[hc]
    and sg_cmds_extra.[hc]
    - add sg_ll_ata_pt() (SATL ATA pass) to sg_cmds_extra.[hc]
  - sg_rdac: fix includes for FreeBSD
  - sg_dd: add 'coe_limit=' option to exit after <n>
    consecutive 'coe' type read errors
  - sgm_dd: print out throughput information when signal arrives
    if time=1 (like sg_dd does already)
  - sg_inq: change '-HHH' so same as '-rr'. Now sg_inq, sg_vpd
    and sdparm output for hdparm with '-HHH'
    -add '-l=<resp_len>' option
  - sg_read_long: add '--pblock' option for physical blocks
  - sg_luns: add '--hex' and '--raw' options
  - sg_requests: add '--hex' and '--raw' options

sg_ident.c
/*
 * Copyright (c) 2005-2007 Douglas Gilbert.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 */

#include <unistd.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>

#include "sg_lib.h"
#include "sg_cmds_basic.h"
#include "sg_cmds_extra.h"

/* A utility program for the Linux OS SCSI subsystem.
 *
 *
 * This program issues these SCSI commands: REPORT IDENTIFYING INFORMATION
 * and SET IDENTIFYING INFORMATION. These commands were called REPORT
 * DEVICE IDENTIFIER and SET DEVICE IDENTIFIER prior to spc4r07.
 */

static char * version_str = "1.06 20070419";

#define ME "sg_ident: "

#define REPORT_ID_INFO_SANITY_LEN 512


static struct option long_options[] = {
        {"ascii", 0, 0, 'A'},
        {"clear", 0, 0, 'C'},
        {"help", 0, 0, 'h'},
        {"itype", 1, 0, 'i'},
        {"raw", 0, 0, 'r'},
        {"set", 0, 0, 'S'},
        {"verbose", 0, 0, 'v'},
        {"version", 0, 0, 'V'},
        {0, 0, 0, 0},
};

static void decode_ii(const unsigned char * iip, int ii_len, int itype,
                      int ascii, int raw, int verbose)
{
    int k, n;

    if (raw) {
        if (ii_len > 0)
            n = fwrite(iip, 1, ii_len, stdout);
        return;
    }
    if (0x7f == itype) {  /* list of available information types */
        for (k = 0; k < (ii_len - 3); k += 4)
            printf("  Information type: %d, Maximum information length: "
                   "%d bytes\n", iip[k], ((iip[k + 2] << 8) + iip[k + 3]));
    } else {        /* single element */
        if (verbose)
            printf("Information:\n");
        if (ii_len > 0) {
            if (ascii)
                printf("%.*s\n", ii_len, (const char *)iip);
            else
                dStrHex((const char *)iip, ii_len, 0); 
        }
    }
}

static void usage()
{
    fprintf(stderr, "Usage: "
          "sg_ident   [--ascii] [--clear] [--help] [--itype=IT] [--raw] "
          "[--set]\n"
          "                  [--verbose] [--version] DEVICE\n"
          "  where:\n"
          "    --ascii|-A      report identifying information as ASCII "
          "(or UTF8) string\n"
          "    --clear|-C      clear (set to zero length) identifying "
          "information\n"
          "    --help|-h       print out usage message\n"
          "    --itype=IT|-i IT    specify information type\n"
          "    --raw|-r        output identifying information to "
          "stdout\n"
          "    --set|-S        invoke set identifying information with "
          "data from stdin\n"
          "    --verbose|-v    increase verbosity of output\n"
          "    --version|-V    print version string and exit\n\n"
          "Performs a SCSI REPORT (or SET) IDENTIFYING INFORMATION command\n"
          );
}

int main(int argc, char * argv[])
{
    int sg_fd, res, c, ii_len;
    unsigned char rdi_buff[REPORT_ID_INFO_SANITY_LEN + 4];
    unsigned char * ucp = NULL;
    int ascii = 0;
    int do_clear = 0;
    int itype = 0;
    int raw = 0;
    int do_set = 0;
    int verbose = 0;
    char device_name[512];
    int ret = 0;

    memset(device_name, 0, sizeof device_name);
    while (1) {
        int option_index = 0;

        c = getopt_long(argc, argv, "AChi:rSvV", long_options,
                        &option_index);
        if (c == -1)
            break;

        switch (c) {
        case 'A':
            ascii = 1;
            break;
        case 'C':
            do_clear = 1;
            break;
        case 'h':
        case '?':
            usage();
            return 0;
        case 'i':
           itype = sg_get_num(optarg);
           if ((itype < 0) || (itype > 127)) {
                fprintf(stderr, "argument to '--itype' should be in range "
                        "0 to 127\n");
                return SG_LIB_SYNTAX_ERROR;
            }
            break;
        case 'r':
            raw = 1;
            break;
        case 'S':
            do_set = 1;
            break;
        case 'v':
            ++verbose;
            break;
        case 'V':
            fprintf(stderr, ME "version: %s\n", version_str);
            return 0;
        default:
            fprintf(stderr, "unrecognised option code 0x%x ??\n", c);
            usage();
            return SG_LIB_SYNTAX_ERROR;
        }
    }
    if (optind < argc) {
        if ('\0' == device_name[0]) {
            strncpy(device_name, argv[optind], sizeof(device_name) - 1);
            device_name[sizeof(device_name) - 1] = '\0';
            ++optind;
        }
        if (optind < argc) {
            for (; optind < argc; ++optind)
                fprintf(stderr, "Unexpected extra argument: %s\n",
                        argv[optind]);
            usage();
            return SG_LIB_SYNTAX_ERROR;
        }
    }

    if (0 == device_name[0]) {
        fprintf(stderr, "missing device name!\n");
        usage();
        return SG_LIB_SYNTAX_ERROR;
    }
    if (do_set && do_clear) {
        fprintf(stderr, "only one of '--clear' and '--set' can be given\n");
        usage();
        return SG_LIB_SYNTAX_ERROR;
    }
    if (ascii && raw) {
        fprintf(stderr, "only one of '--ascii' and '--raw' can be given\n");
        usage();
        return SG_LIB_SYNTAX_ERROR;
    }
    if ((do_set || do_clear) && (raw || ascii)) {
        fprintf(stderr, "'--set' cannot be used with either '--ascii' or "
                "'--raw'\n");
        usage();
        return SG_LIB_SYNTAX_ERROR;
    }
    sg_fd = sg_cmds_open_device(device_name, 0 /* rw */, verbose);
    if (sg_fd < 0) {
        fprintf(stderr, ME "open error: %s: %s\n", device_name,
                safe_strerror(-sg_fd));
        return SG_LIB_FILE_ERROR;
    }

    memset(rdi_buff, 0x0, sizeof(rdi_buff));
    if (do_set || do_clear) {
        if (do_set) {
            res = fread(rdi_buff, 1, REPORT_ID_INFO_SANITY_LEN + 2, stdin); 
            if (res <= 0) {
                fprintf(stderr, "no data read from stdin; to clear "
                        "identifying information use '--clear' instead\n");
                ret = -1;
                goto err_out;
            } else if (res > REPORT_ID_INFO_SANITY_LEN) {
                fprintf(stderr, "SPC-4 limits information length to 512 "
                        "bytes\n");
                ret = -1;
                goto err_out;
            }
            ii_len = res;
            res = sg_ll_set_id_info(sg_fd, itype, rdi_buff, ii_len, 1,
                                    verbose);
        } else    /* do_clear */
            res = sg_ll_set_id_info(sg_fd, itype, rdi_buff, 0, 1, verbose);
        if (res) {
            ret = res;
            if (SG_LIB_CAT_NOT_READY == res)
                fprintf(stderr, "Set identifying information command, device "
                        "not ready\n");
            else if (SG_LIB_CAT_INVALID_OP == res)
                fprintf(stderr, "Set identifying information command not "
                        "supported\n");
            else if (SG_LIB_CAT_UNIT_ATTENTION == res)
                fprintf(stderr, "Set identifying information, unit "
                        "attention\n");
            else if (SG_LIB_CAT_ABORTED_COMMAND == res)
                fprintf(stderr, "Set identifying information, aborted "
                        "command\n");
            else if (SG_LIB_CAT_ILLEGAL_REQ == res)
                fprintf(stderr, "bad field in Set identifying information "
                        "cdb including unsupported service action\n");
            else {
                fprintf(stderr, "Set identifying information command "
                        "failed\n");
                if (0 == verbose)
                    fprintf(stderr, "    try '-v' for more information\n");
            }
        }
    } else {    /* do report identifying information */
        res = sg_ll_report_id_info(sg_fd, itype, rdi_buff, 4, 1, verbose);
        if (0 == res) {
            ii_len = (rdi_buff[0] << 24) + (rdi_buff[1] << 16) + 
                         (rdi_buff[2] << 8) + rdi_buff[3];
            if ((! raw) && (verbose > 0))
                printf("Reported identifying information length = %d\n",
                       ii_len);
            if (0 == ii_len) {
                if (verbose > 1)
                    fprintf(stderr, "    This implies the device has an "
                            "empty information field\n");
                goto err_out;
            }
            if (ii_len > REPORT_ID_INFO_SANITY_LEN) {
                fprintf(stderr, "    That length (%d) seems too long for an "
                        "information\n", ii_len);
                ret = -1;
                goto err_out;
            }
            ucp = rdi_buff;
            res = sg_ll_report_id_info(sg_fd, itype, ucp, ii_len + 4, 1,
                                       verbose);
            if (0 == res) {
                ii_len = (ucp[0] << 24) + (ucp[1] << 16) + (ucp[2] << 8) +
                         ucp[3];
                decode_ii(ucp + 4, ii_len, itype, ascii, raw, verbose);
            } else
                ret = res;
        } else
            ret = res;
        if (ret) {
            if (SG_LIB_CAT_NOT_READY == ret)
                fprintf(stderr, "Report identifying information command, "
                        "device not ready\n");
            else if (SG_LIB_CAT_UNIT_ATTENTION == ret)
                fprintf(stderr, "Report identifying information, unit "
                        "attention\n");
            else if (SG_LIB_CAT_ABORTED_COMMAND == ret)
                fprintf(stderr, "Report identifying information, aborted "
                        "command\n");
            else if (SG_LIB_CAT_INVALID_OP == ret)
                fprintf(stderr, "Report identifying information command "
                        "not supported\n");
            else if (SG_LIB_CAT_ILLEGAL_REQ == ret)
                fprintf(stderr, "bad field in Report identifying "
                        "information cdb including unsupported service "
                        "action\n");
            else {
                fprintf(stderr, "Report identifying information command "
                        "failed\n");
                if (0 == verbose)
                    fprintf(stderr, "    try '-v' for more "
                            "information\n");
            }
        }
    }

err_out:
    res = sg_cmds_close_device(sg_fd);
    if (res < 0) {
        fprintf(stderr, "close error: %s\n", safe_strerror(-res));
        if (0 == ret)
            return SG_LIB_FILE_ERROR;
    }
    return (ret >= 0) ? ret : SG_LIB_CAT_OTHER;
}
